[DDP] fix param to name mapping

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#57771 [DDP] fix param to name mapping**

This mapping didn't work properly when certain parameters didn't
require grad. Fixed that and added a test.

Differential Revision: [D28265636](https://our.internmc.facebook.com/intern/diff/D28265636/)

rohan-varma committed:
[DDP] fix param to name mapping

This mapping didn't work properly when certain parameters didn't
require grad. Fixed that and added a test.

Differential Revision: [D28265636](https://our.internmc.facebook.com/intern/diff/D28265636/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] fix param to name mapping"


This mapping didn't work properly when certain parameters didn't
require grad. Fixed that and added a test.

Differential Revision: [D28265636](https://our.internmc.facebook.com/intern/diff/D28265636/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] fix param to name mapping"


This mapping didn't work properly when certain parameters didn't
require grad. Fixed that and added a test.

Differential Revision: [D28265636](https://our.internmc.facebook.com/intern/diff/D28265636/)

[ghstack-poisoned]

SciPioneer commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -6049,6 +6049,30 @@ def test_ddp_build_param_to_name_mapping(self):
+                    self.lin = nn.Linear(10, 10)
+                    # Is not tracked by DDP and should not show up in param to
+                    # name mapping.
+                    self.lin.bias.requires_grad_(False)
```
Shall we add another test that validate the `ValueError` if `requires_grad_` is false?

rohan-varma replied:
That ValueError is kind of internal to DDP, it would be a bug in DDP if user could construct inputs that trigger that error. For example, that error would be triggered before this requires_grad fix. Will add a line to those errors that it is a bug and to report an issue.

rohan-varma replied:
Actually, I guess we can test these error messages by manually changing parameters in this test case and above, so might as well add that.

rohan-varma committed:
Update on "[DDP] fix param to name mapping"


This mapping didn't work properly when certain parameters didn't
require grad. Fixed that and added a test.

Differential Revision: [D28265636](https://our.internmc.facebook.com/intern/diff/D28265636/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] fix param to name mapping"


This mapping didn't work properly when certain parameters didn't
require grad. Fixed that and added a test.

Differential Revision: [D28265636](https://our.internmc.facebook.com/intern/diff/D28265636/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[DDP] fix param to name mapping"


This mapping didn't work properly when certain parameters didn't
require grad. Fixed that and added a test.

Differential Revision: [D28265636](https://our.internmc.facebook.com/intern/diff/D28265636/)

[ghstack-poisoned]
